### DISCUSSION TITLE
How can I use custom Certificate for a particular KafkaUser?

### DISCUSSION THREAD

**Discussion by Mehmoodhassan** [strimzi/strimzi-kafka-operator#9138] [Q&A]
Hi everyone,
I've deployed the Strimzi Kafka Operator in my Azure Kubernetes Service (AKS) cluster, and it's working smoothly. By default, when I create a KafkaUser, the operator automatically generates an associated secret that includes the user's certificate (`user.crt`) and private key (`user.key`). However, I'd like to customize this process by providing my own certificates and keys instead of relying on the automatic generation of these credentials. 
How can I do this? 
Would be happy if you share a example kafkaUser yaml file. 
@scholzj  will appreciate your answer.

**Comment by scholzj** [comment]
You should check the documentation about the Custom Clients CA in the `Kafka` resource / Kafka cluster and about the `tls-external` authentication type in the `KafkaUser` resources

**Reply by Mehmoodhassan** [reply]
@scholzj Thanks for your prompt response :) Yeah I go through the docs https://strimzi.io/docs/operators/in-development/configuring.html#type-KafkaUserTlsExternalClientAuthentication-reference but somehow due to less knowledge about kafka don't know in KafkaUser defination where I need to specify the secret with custom certificate and also can't see any example there. I would be happy if you link any example here. Thanks in advance.

**Reply by scholzj** [reply]
You don't specify any secret. If you issued your own certificate, then you already have it and you can do whatever you want with it. But as I said, you also need to configure the Custom Clients CA -> that is what establishes the trust in the brokers.

**Reply by Mehmoodhassan** [reply]
Thanks again @scholzj Actually I want to setup both below options at a same time 

- Custom CAs for external communication purpose
- let the kakfa to create CAs for some specific users

Is that possible?

**Reply by scholzj** [reply]
If it is the same CA, that is fine. If those are two different, I'm not sure that will work. We really support only one CA for the user authentication.

**Reply by Mehmoodhassan** [reply]
Thanks @scholzj One last question. when I will create a user with `authentication: tls-external` to use my custom Certificate. Do I need to pass custom certificate somewhere else or only just need to create a secret and label/annotate it? 
I mean regarding below properties because I don't want to use custom CA for cluster and client just want to have one for a kafkaUser.
```
 clusterCa:
    generateCertificateAuthority: false
  clientsCa:
    generateCertificateAuthority: false
```

One more thing I am facing a really weird behavior as soon as I label the custom secret it automatically get deleted :( I don't why? Am I missing something?

**Reply by scholzj** [reply]
You do not need to pass the certificate anywhere nor create the Secret. The certificate is authenticated through the CA. So it has to be signed by the Clients CA. Whether it is the Strimzi one or your custom one does not really matter. But it is trusted through the CA. So you cannot sign ti with some random CA.

**Reply by Mehmoodhassan** [reply]
Thanks for the quick response @scholzj . Now I have a new certificate which is signed by the the Clients CA and I tries to use that CA and KEY with default producer and consumer to connect to the Broker but I am having below error at both Producer and consumer
![image](https://github.com/strimzi/strimzi-kafka-operator/assets/52590381/03ec2fdb-f8e7-47b5-b630-ba6c3262541e)

would be thankful if you leave your useful thoughts here :)

**Reply by scholzj** [reply]
I guess that error suggests some configuration issue. But without knowing how you issues the certificate and how you configured it, it is hard to say anything more.

**Reply by Mehmoodhassan** [reply]
@ Actually I garbed the default Clients CA from  cluster which was created from operator and then try to sign new certificate from that one by below steps: 
### Create a Certificate Signing Request (CSR) for the Certificate to Be Signed:
`openssl req -new -newkey rsa:2048 -keyout certificate-to-sign.key -out certificate-to-sign.csr -subj "/CN=kafka-username"`
### Sign the CSR with the CA Certificate:
`openssl x509 -req -in certificate-to-sign.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out signed-certificate.crt -days 365`

here the ca.crt and ca.key is Clients CA.

if I use a certificate from a kafkauser which is created by operator it works fine.
